File: packages/platform/src/ipc/account.ts
/**
 * 账号管理 IPC 封装 — 多账号 CRUD + 状态切换。
 *
 * 后端：壳层 `commands/account.rs`（InMemoryAccountStore + app::account::AccountService）。
 *
 * @author agent
 * @created 2026-08-13
 */

import { callRequest } from "./invoke";

/** 账号状态（与 Rust `AccountStatus` 对齐）。 */
export type AccountStatus = "active" | "disabled";

/** 账号所属平台。 */
export type AccountPlatform = "xianyu" | "ali1688" | "xiaohongshu";

/** 账号（与 Rust `app::account::XianyuAccount` 对齐的核心字段）。 */
export interface XianyuAccount {
  id: number;
  owner_id: number;
  /** 账号标识（全局唯一）。 */
  account_id: string;
  display_name: string;
  /** 头像 URL（连接后从闲鱼同步）。 */
  avatar_url: string;
  login_id: string;
  login_password: string;
  unb: string;
  /** 本站业务 Cookie（闲鱼账号为 goofish；1688 账号为 1688 袋）。 */
  cookie: string;
  /**
   * 1688 Cookie（兼容旧双站字段；分站登录后可为空）。
   *
   * @author Xiaoman
   * @created 2026-08-22
   */
  cookie_1688?: string;
  /**
   * 平台：`xianyu` / `ali1688`。
   *
   * @author Xiaoman
   * @created 2026-08-22
   */
  platform?: AccountPlatform | string;
  /** qr / password。 */
  login_method: string;
  status: AccountStatus;
  pause_duration_minutes: number;
}

/** 账号更新补丁（缺省字段不更新）。 */
export interface AccountUpdate {
  display_name?: string;
  avatar_url?: string;
  status?: AccountStatus;
  login_id?: string;
  login_password?: string;
  /** 更新 Cookie（扫码重登 / 风控后在真实浏览器获取的新 Cookie）。 */
  cookie?: string;
  /**
   * 更新 1688 Cookie。
   *
   * @author Xiaoman
   * @created 2026-08-22
   */
  cookie_1688?: string;
}

/** 查询账号列表。 */
export function accountList(ownerId: number): Promise<XianyuAccount[]> {
  return callRequest<XianyuAccount[]>("account_list", { ownerId }).then((response) => response.data);
}

/** 新建账号。 */
export function accountCreate(
  ownerId: number,
  account: XianyuAccount,
): Promise<XianyuAccount> {
  return callRequest<XianyuAccount>("account_create", { ownerId, account }).then(
    (response) => response.data,
  );
}

/** 更新账号（部分字段）。 */
export function accountUpdate(
  ownerId: number,
  accountId: string,
  patch: AccountUpdate,
): Promise<XianyuAccount> {
  return callRequest<XianyuAccount>("account_update", { ownerId, accountId, patch }).then(
    (response) => response.data,
  );
}

/** 切换账号启用状态。 */
export function accountSetStatus(
  ownerId: number,
  accountId: string,
  status: AccountStatus,
): Promise<void> {
  return callRequest<void>("account_set_status", {
    request: { owner_id: ownerId, account_id: accountId, status },
  }).then(() => undefined);
}

/** 删除账号。 */
export function accountDelete(ownerId: number, accountId: string): Promise<void> {
  return callRequest<void>("account_delete", {
    request: { owner_id: ownerId, account_id: accountId },
  }).then(() => undefined);
}

/** 探测账号 Cookie 是否仍在线（1688 → 浏览器；闲鱼/小红书 → HTTP 刷新 token）。 */
export function accountProbeLogin(ownerId: number, accountId: string): Promise<boolean> {
  return callRequest<boolean>("account_probe_login", {
    request: { owner_id: ownerId, account_id: accountId },
  }).then((response) => response.data);
}

// ========== 业务账号扫码登录（复用 sidecar 扫码能力，成功后自动创建/更新账号） ==========

/** 扫码启动响应（与 Rust `ChannelIpcQrStartResponse` 对齐）。 */
export interface AccountQrStartResult {
  ok: boolean;
  status: string;
  session_id: string | null;
  qr_base64: string | null;
  detail: string | null;
}

/** 扫码状态轮询响应（与 Rust `ChannelIpcQrCheckResponse` 对齐）。 */
export interface AccountQrCheckResult {
  ok: boolean;
  status: string;
  session_id: string | null;
  cookies: unknown[] | null;
  detail: string | null;
  qr_base64: string | null;
}

/** 启动业务账号扫码登录。
 *
 * @author Xiaoman
 * @created 2026-08-22
 *
 * @param name - 可选展示名
 * @param platform - `xianyu` / `ali1688`
 */
export function accountQrStart(
  name?: string,
  platform: AccountPlatform = "xianyu",
): Promise<AccountQrStartResult> {
  return callRequest<AccountQrStartResult>("account_qr_start", {
    request: { name: name ?? null, platform },
  }).then((response) => response.data);
}

/** 轮询扫码状态（成功后账号已自动创建/更新）。
 *
 * @author Xiaoman
 * @created 2026-08-22
 *
 * @param sessionId - 扫码会话 ID
 * @param platform - 须与 start 一致
 */
export function accountQrCheck(
  sessionId: string,
  platform: AccountPlatform = "xianyu",
): Promise<AccountQrCheckResult> {
  return callRequest<AccountQrCheckResult>("account_qr_check", {
    request: { session_id: sessionId, platform },
  }).then((response) => response.data);
}

/** 取消业务账号扫码登录。
 *
 * @author Xiaoman
 * @created 2026-08-22
 *
 * @param sessionId - 扫码会话 ID
 * @param platform - 须与 start 一致
 */
export function accountQrCancel(
  sessionId: string,
  platform: AccountPlatform = "xianyu",
): Promise<void> {
  return callRequest<void>("account_qr_cancel", {
    request: { session_id: sessionId, platform },
  }).then(() => undefined);
}

// ========== 业务账号渠道连接（扫码后自动连接 / 手动断开连接） ==========

/** 连接业务账号（建立渠道 websocket 设备监听）。 */
export function accountConnect(ownerId: number, accountId: string): Promise<string> {
  return callRequest<string>("account_connect", {
    request: { owner_id: ownerId, account_id: accountId },
  }).then((response) => response.data);
}

/** 断开业务账号的渠道连接。 */
export function accountDisconnect(ownerId: number, accountId: string): Promise<void> {
  return callRequest<void>("account_disconnect", {
    request: { owner_id: ownerId, account_id: accountId },
  }).then(() => undefined);
}

/** 查询业务账号的渠道连接状态（connected / connecting / disconnected / error）。 */
export function accountConnectionState(ownerId: number, accountId: string): Promise<string> {
  return callRequest<string>("account_connection_state", {
    request: { owner_id: ownerId, account_id: accountId },
  }).then((response) => response.data);
}

File: apps/desktop/src-tauri/src/commands/account.rs
//! 账号业务 IPC（CRUD / 扫码 / 登录探针）。

pub use crud::{
    account_create, account_delete, account_list, account_probe_login, account_set_status,
    account_update, AccountHandle,
};
pub use qr::{
    account_qr_cancel, account_qr_check, account_qr_start, AccountQrHandle, PostQrLoginHook,
};

mod crud {
    // 账号管理 Tauri commands — 多账号 CRUD + 状态切换。
    //
    // 壳层组合：`InMemoryAccountStore` → `crate::domain::account::AccountService`（校验 + 编排）。

    use crate::application::account::probe_account_session;
    use crate::bootstrap::state::AppState;
    use crate::commands::IpcResponse;
    use crate::domain::account::{AccountService, AccountStatus, AccountUpdate, XianyuAccount};
    use crate::infrastructure::database::stores::InMemoryAccountStore;
    use serde::Deserialize;
    use std::sync::Arc;
    use tauri::State;

    /// 账号状态变更入参。
    #[derive(Debug, Deserialize)]
    pub struct AccountStatusRequest {
        pub owner_id: i64,
        pub account_id: String,
        pub status: String,
    }

    /// 账号删除入参。
    #[derive(Debug, Deserialize)]
    pub struct AccountDeleteRequest {
        pub owner_id: i64,
        pub account_id: String,
    }

    /// 账号探针入参（owner + account_id）。
    #[derive(Debug, Deserialize)]
    pub struct AccountProbeRequest {
        pub owner_id: i64,
        pub account_id: String,
    }

    /// 账号服务句柄（setup 时注册到 Tauri 状态）。
    pub struct AccountHandle {
        pub store: Arc<InMemoryAccountStore>,
    }

    /// 查询账号列表。
    #[tauri::command]
    pub fn account_list(
        state: State<'_, AccountHandle>,
        owner_id: i64,
    ) -> crate::contracts::DingDaResult<IpcResponse<Vec<XianyuAccount>>> {
        let service = AccountService::new(state.store.as_ref());
        let result = service
            .list(owner_id)
            .map_err(crate::contracts::DingDaError::wrap)?;
        Ok(IpcResponse::ok(result))
    }

    /// 新建账号（含归属/唯一性校验）。
    #[tauri::command]
    pub fn account_create(
        state: State<'_, AccountHandle>,
        owner_id: i64,
        account: XianyuAccount,
    ) -> crate::contracts::DingDaResult<IpcResponse<XianyuAccount>> {
        let service = AccountService::new(state.store.as_ref());
        let result = service
            .create(owner_id, &account)
            .map_err(crate::contracts::DingDaError::wrap)?;
        Ok(IpcResponse::ok(result))
    }

    /// 更新账号（部分字段补丁）。
    #[tauri::command]
    pub fn account_update(
        state: State<'_, AccountHandle>,
        owner_id: i64,
        account_id: String,
        patch: AccountUpdate,
    ) -> crate::contracts::DingDaResult<IpcResponse<XianyuAccount>> {
        let service = AccountService::new(state.store.as_ref());
        let result = service
            .update(owner_id, &account_id, &patch)
            .map_err(crate::contracts::DingDaError::wrap)?;
        Ok(IpcResponse::ok(result))
    }

    /// 切换账号启用状态。
    #[tauri::command]
    pub fn account_set_status(
        state: State<'_, AccountHandle>,
        request: AccountStatusRequest,
    ) -> crate::contracts::DingDaResult<IpcResponse<()>> {
        let service = AccountService::new(state.store.as_ref());
        let status = AccountStatus::from_str(&request.status);
        service
            .set_status(request.owner_id, &request.account_id, status)
            .map_err(crate::contracts::DingDaError::wrap)?;
        Ok(IpcResponse::ok(()))
    }

    /// 删除账号（归属校验）。
    #[tauri::command]
    pub fn account_delete(
        state: State<'_, AccountHandle>,
        request: AccountDeleteRequest,
    ) -> crate::contracts::DingDaResult<IpcResponse<()>> {
        let service = AccountService::new(state.store.as_ref());
        service
            .delete(request.owner_id, &request.account_id)
            .map_err(crate::contracts::DingDaError::wrap)?;
        Ok(IpcResponse::ok(()))
    }

    /// 探测账号 Cookie 是否仍在线（1688 → 浏览器探针；闲鱼/小红书 → HTTP 刷新 token）。
    #[tauri::command]
    pub async fn account_probe_login(
        state: State<'_, AccountHandle>,
        app_state: State<'_, AppState>,
        request: AccountProbeRequest,
    ) -> crate::contracts::DingDaResult<IpcResponse<bool>> {
        let online = probe_account_session(
            state.store.as_ref(),
            app_state.lifecycle.as_ref(),
            request.owner_id,
            &request.account_id,
        )
        .await;
        Ok(IpcResponse::ok(online))
    }
}

mod qr {
    // 业务账号扫码登录 Tauri commands — 按平台（闲鱼 / 1688）扫码创建账号。
    //
    // 复用 sidecar 的 `channel_qr_*`；`platform` 决定登录页与 Cookie 落袋。
    // 落库成功后的平台后置逻辑（闲鱼建渠道 WS）由各平台 bootstrap 注入，
    // 本模块不做任何平台分支。
    //
    // 作者：Xiaoman
    // 创建时间：2026-08-20

    use crate::application::channel::dispatcher::ChannelDispatcher;
    use crate::bootstrap::state::AppState;
    use crate::commands::IpcResponse;
    use crate::contracts::{
        ChannelIpcQrCancelResponse, ChannelIpcQrCheckResponse, ChannelIpcQrStartResponse,
        ChannelSidecarQrCancelRequest, ChannelSidecarQrCheckRequest, ChannelSidecarQrStartRequest,
    };
    use crate::domain::account::{
        AccountService, AccountStore, AccountUpdate, LoginMethod, XianyuAccount,
    };
    use crate::infrastructure::database::account::normalize_account_platform;
    use crate::infrastructure::database::account_qr::account_from_cookies;
    use crate::infrastructure::database::stores::InMemoryAccountStore;
    use serde::Deserialize;
    use std::future::Future;
    use std::pin::Pin;
    use std::sync::{Arc, RwLock};
    use tauri::{AppHandle, Manager, State};
    use tracing::info;

    /// 扫码落库成功后的平台后置逻辑（闲鱼自动建渠道 WS 等）。
    ///
    /// 由各平台 bootstrap 注入：`xianyu` 传连接钩子，`ali1688` 传 `None`。
    ///
    /// 参数依次为调度器、业务账号存储、归属用户 id、已落库的业务账号；
    /// 返回统一的业务结果，失败会上抛给调用方。
    ///
    /// 作者：Xiaoman
    /// 创建时间：2026-08-22
    pub type PostQrLoginHook = Arc<
        dyn Fn(
                Arc<ChannelDispatcher>,
                Arc<InMemoryAccountStore>,
                i64,
                XianyuAccount,
            )
                -> Pin<Box<dyn Future<Output = crate::contracts::DingDaResult<()>> + Send>>
            + Send
            + Sync,
    >;

    /// 业务账号扫码服务句柄（setup 时注册到 Tauri 状态）。
    ///
    /// 作者：Xiaoman
    /// 创建时间：2026-08-20
    pub struct AccountQrHandle {
        pub store: Arc<InMemoryAccountStore>,
        /// 扫码成功后的平台后置逻辑；`None` 表示无后置动作（1688 仅落库）。
        /// 由两站共用 `core::bootstrap` 初始化为 `None`，闲鱼 bootstrap 启动时写入。
        pub post_login: RwLock<Option<PostQrLoginHook>>,
    }

    /// 扫码启动入参。
    ///
    /// 作者：Xiaoman
    /// 创建时间：2026-08-20
    #[derive(Debug, Deserialize)]
    pub struct AccountQrStartRequest {
        /// 展示名称（可选）。
        #[serde(default)]
        pub name: Option<String>,
        /// 平台：`xianyu` / `ali1688`；缺省闲鱼。
        #[serde(default)]
        pub platform: Option<String>,
    }

    /// 扫码轮询入参（带平台，避免与会话平台不匹配）。
    ///
    /// 作者：Xiaoman
    /// 创建时间：2026-08-22
    #[derive(Debug, Deserialize)]
    pub struct AccountQrSessionRequest {
        pub session_id: String,
        #[serde(default)]
        pub platform: Option<String>,
    }

    /// 启动业务账号扫码登录。
    ///
    /// 作者：Xiaoman
    /// 创建时间：2026-08-20
    #[tauri::command]
    pub async fn account_qr_start(
        state: State<'_, AppState>,
        request: AccountQrStartRequest,
    ) -> crate::contracts::DingDaResult<IpcResponse<ChannelIpcQrStartResponse>> {
        state
            .license
            .ensure_licensed()
            .await
            .map_err(crate::contracts::DingDaError::wrap)?;

        let platform = normalize_account_platform(request.platform.as_deref().unwrap_or("xianyu"));
        let sidecar_request = ChannelSidecarQrStartRequest {
            account_id: String::new(),
            trace_id: Some(
                request
                    .name
                    .clone()
                    .unwrap_or_else(|| format!("account-qr-{platform}")),
            ),
            platform: Some(platform.to_string()),
        };
        let sidecar = state.lifecycle.client();
        let response =
            crate::infrastructure::sidecar::channel_login::qr_start(sidecar, sidecar_request)
                .await
                .map_err(crate::contracts::DingDaError::wrap)?;

        Ok(IpcResponse::ok(ChannelIpcQrStartResponse {
            ok: response.ok,
            status: response.status,
            session_id: response.session_id,
            qr_base64: response.qr_base64,
            detail: response.detail,
        }))
    }

    /// 轮询扫码状态；成功后按平台落库。
    ///
    /// 作者：Xiaoman
    /// 创建时间：2026-08-20
    #[tauri::command]
    pub async fn account_qr_check(
        state: State<'_, AppState>,
        app: AppHandle,
        dispatcher: State<'_, Arc<ChannelDispatcher>>,
        request: AccountQrSessionRequest,
    ) -> crate::contracts::DingDaResult<IpcResponse<ChannelIpcQrCheckResponse>> {
        state
            .license
            .ensure_licensed()
            .await
            .map_err(crate::contracts::DingDaError::wrap)?;

        let platform = normalize_account_platform(request.platform.as_deref().unwrap_or("xianyu"));
        let sidecar_request = ChannelSidecarQrCheckRequest {
            session_id: request.session_id.clone(),
            trace_id: Some(request.session_id.clone()),
            platform: Some(platform.to_string()),
        };
        let sidecar = state.lifecycle.client();
        let response =
            crate::infrastructure::sidecar::channel_login::qr_check(sidecar, sidecar_request)
                .await
                .map_err(crate::contracts::DingDaError::wrap)?;

        if response.status == "success" {
            if let Some(cookies) = response.cookies.clone() {
                let account = account_from_cookies(platform, &cookies);
                let handle = app.state::<AccountQrHandle>();
                let service = AccountService::new(handle.store.as_ref());

                match handle.store.get_account(1, &account.account_id) {
                    Ok(Some(_)) => {
                        service
                            .update(
                                1,
                                &account.account_id,
                                &AccountUpdate {
                                    cookie: Some(account.cookie.clone()),
                                    cookie_1688: Some(account.cookie_1688.clone()),
                                    unb: Some(account.unb.clone()),
                                    platform: Some(platform.to_string()),
                                    login_method: Some(LoginMethod::Qr),
                                    last_login_at: Some(now_string()),
                                    ..Default::default()
                                },
                            )
                            .map_err(crate::contracts::DingDaError::wrap)?;
                    }
                    _ => {
                        service
                            .create(1, &account)
                            .map_err(crate::contracts::DingDaError::wrap)?;
                    }
                }

                info!(
                    account = %account.account_id,
                    platform,
                    "扫码成功，账号已落库"
                );

                // 平台后置逻辑（闲鱼自动建渠道 WS / 拉资料）由 bootstrap 写入；
                // 1688 无后置，`post_login` 为 `None`，此处不执行。
                let post_login = handle
                    .post_login
                    .read()
                    .unwrap_or_else(|poisoned| poisoned.into_inner())
                    .clone();
                if let Some(hook) = post_login {
                    hook(dispatcher.inner().clone(), handle.store.clone(), 1, account)
                        .await
                        .map_err(crate::contracts::DingDaError::wrap)?;
                }
            }
        }

        Ok(IpcResponse::ok(ChannelIpcQrCheckResponse {
            ok: response.ok,
            status: response.status,
            session_id: response.session_id,
            cookies: response.cookies,
            detail: response.detail,
            qr_base64: response.qr_base64,
        }))
    }

    /// 取消业务账号扫码登录。
    ///
    /// 作者：Xiaoman
    /// 创建时间：2026-08-20
    #[tauri::command]
    pub async fn account_qr_cancel(
        state: State<'_, AppState>,
        request: AccountQrSessionRequest,
    ) -> crate::contracts::DingDaResult<IpcResponse<ChannelIpcQrCancelResponse>> {
        state
            .license
            .ensure_licensed()
            .await
            .map_err(crate::contracts::DingDaError::wrap)?;

        let platform = normalize_account_platform(request.platform.as_deref().unwrap_or("xianyu"));
        let sidecar_request = ChannelSidecarQrCancelRequest {
            session_id: request.session_id.clone(),
            trace_id: Some(request.session_id.clone()),
            platform: Some(platform.to_string()),
        };
        let sidecar = state.lifecycle.client();
        let response =
            crate::infrastructure::sidecar::channel_login::qr_cancel(sidecar, sidecar_request)
                .await
                .map_err(crate::contracts::DingDaError::wrap)?;

        Ok(IpcResponse::ok(ChannelIpcQrCancelResponse {
            ok: response.ok,
            detail: response.detail,
        }))
    }

    fn now_string() -> String {
        use chrono::Utc;
        Utc::now().format("%Y-%m-%d %H:%M:%S").to_string()
    }
}
